### ISSUE TITLE
fine-tune poorly on my own dataset

### ISSUE THREAD

**Issue by XinnWang** [TRI-ML/packnet-sfm#41]
Hi, thanks for the great work ! I am trying to fine-tune the 'PackNet01_MR_selfsup_K.ckpt' on my own dataset, but the performance is not good. Here are some examples:

If i use 'PackNet01_MR_selfsup_K.ckpt' without fine-tune, the result looks like this:
![image](https://user-images.githubusercontent.com/5515327/86346085-ffba5180-bc8e-11ea-9c23-b6f9c9b4c1ac.png)

after fine-tune this model, i get this result:
![image](https://user-images.githubusercontent.com/5515327/86346176-1f517a00-bc8f-11ea-8aa2-ebf2613c4ba2.png)

The following is the yaml file i used for training.
```
arch:
    max_epochs: 10
checkpoint:
    filepath: /workspace/packnet-sfm/results/chept
    save_top_k: -1
model:
    name: 'SelfSupModel'
    checkpoint_path: /data/models/PackNet01_MR_semisup_CStoK.ckpt
    optimizer:
        name: 'Adam'
        depth:
            lr: 0.0002
        pose:
            lr: 0.0002
    scheduler:
        name: 'StepLR'
        step_size: 30
        gamma: 0.5
    depth_net:
        name: 'PackNet01'
        version: '1A'
    pose_net:
        name: 'PoseNet'
        version: ''
    params:
        crop: 'garg'
        min_depth: 0.0
        max_depth: 80.0
datasets:
    augmentation:
        image_shape: (192, 640)
    train:
        batch_size: 1
        dataset: ['Image']
        path: ['my_training_dataset']
        split: ['{:03d}']
        repeat: [1]
    validation:
        dataset: ['Image']
        path: ['my_eval_dataset']
        split: ['{:03d}']
    test:
        dataset: ['Image']
        path: ['my_test_dataset']
        split: ['{:03d}']
```

i also changed the dummy_calibration parames inside datasets/image_dataset.py:
```
def dummy_calibration(image):
    w, h = [float(d) for d in image.size]
    return np.array([[343.169585 , 0.    , 321.358181],
                     [0.    , 344.619087, 93.813611],
                     [0.    , 0.    , 1.          ]])
```

My dataset contains 3k images, resolution is (192,640). (only mono images)
The training is done on 4 x Tesla M40, with NVIDIA-SMI 418.56 Driver Version: 418.56 CUDA Version: 10.1
After the training, the 'Avg.Loss' drops from 0.0370 to 0.0350.  (it drops really slow.)

I know my training epochs is too few. i am still training it now. but the Avg.Loss is always around 0.0350. 
Is there any other possible reasons about this poor result? Any help or suggestion will be appreciated. Thanks !

**Comment by VitorGuizilini-TRI** [comment]
Can you try a smaller learning rate, since you are fine-tuning? And also evaluate on the training dataset, maybe it's overfitting to this smaller dataset and poorly transferring to the evaluation/test splits.  Other than that, the dataset could be ill-fitted for self-supervised learning (low-resolution images, distortions, dynamic objects, not enough motion).

**Comment by XinnWang** [comment]
Thanks for your reply and suggestions. i find the poor performance is caused by 'not enough motion'. After i delete some interval frames, the result looks better. So, what is the best motion for self-supervised learning ? For example, consider a near vehicle in the left side of image, its pixels will move backward between two frames. How much the moving distance is necessary for a good self-supervised learning ? What is the relationship between this distance and the 'max_depth' in the yaml file ? Thanks.

**Comment by VitorGuizilini-TRI** [comment]
It's tough to say, we need enough motion so there is a baseline for depth estimation, but not too much so there is enough overlap between the two images. A proper evaluation of different datasets and why self-supervision works or doesn't work on them (and ways to make it work) would be a very good study!

On the yaml file, max_depth just determines the range in which the depth metrics will be evaluated, it doesn't affect training. But I see what you mean, maybe if you care about longer ranges you need more motion to see enough relative displacement. If it's an indoors dataset, with smaller ranges, you might not need that much for self-supervision to kick in and start working.

**Comment by XinnWang** [comment]
I think maybe the static obstacles ( vehicles, tress ) are also crucial for self-supervised training ( especially near obstacles ). In one of my datasets, the road is wide and no other obstacles in the images, training performance is not good. Like you said, a proper evaluation of datasets would be a good study.

Now my training result looks like:
![color_depth5](https://user-images.githubusercontent.com/5515327/86555675-9df63200-bf83-11ea-911a-67dfad77a79c.png)
i also reconstruct its 3d point cloud like this:
![Screenshot from 2020-07-06 12-23-13](https://user-images.githubusercontent.com/5515327/86555732-c847ef80-bf83-11ea-8c49-d5ef18163ac5.png)

As you can see, the traffic line is good, but the vehicle is not good. I'm thinking if the training can be supervised by vehicle speed, the result can be better ? (i have the vehicle speed in the dataset, but i don't have lidar sensors.). Since only self-supervised training is provided in the example, can you provide more instructions for semi-supervised training ? Thanks.

**Comment by VitorGuizilini-TRI** [comment]
If that's the case you can use the VelSupModel, it supervises the pose network with velocity estimates, so you have scale-aware models. The ImageDataset does not support that, though, so you have to modify it to also load speed information. Another way is use it to filter static frames, so instead of doing it manually you can set a threshold on minimum/maximum speed, and discard frames where the vehicle is moving slower or faster than these values.

**Comment by XinnWang** [comment]
Hi, i try to modify the ImageDataset to load speed information. i read your code and find that the 'speed information' is the relative motion between two frames, so i just load the motion (which is a 4x4 matrix) for each image, the code is like this:

```
sample = {
            'idx': idx,
            'filename': '%s_%s' % (session, os.path.splitext(filename)[0]),
            'rgb': image,
            'intrinsics': dummy_calibration(image),
            'pose_context': ["relative motion between this frame and its previous frame"]
        }
```

is this correct ? how is the coordinate system of the relative motion ? Thanks.

**Comment by XinnWang** [comment]
Hi, i find that i was wrong. i changed the code like this:

```
sample = {
            'idx': idx,
            'filename': '%s_%s' % (session, os.path.splitext(filename)[0]),
            'rgb': image,
            'intrinsics': dummy_calibration(image),
            'pose': ["absolute pose for the current frame"]
            'pose_context': ["relative motion between this current pose and context poses"]
        }
```

i guess this should be right. : )

**Comment by surfii3z** [comment]
> Hi, i find that i was wrong. i changed the code like this:
> 
> ```
> sample = {
>             'idx': idx,
>             'filename': '%s_%s' % (session, os.path.splitext(filename)[0]),
>             'rgb': image,
>             'intrinsics': dummy_calibration(image),
>             'pose': ["absolute pose for the current frame"]
>             'pose_context': ["relative motion between this current pose and context poses"]
>         }
> ```
> 
> i guess this should be right. : )

Hi, I am also trying to train the network with VelSupModel. 

Could you please share the result if it's working well or not?

Thanks
